### PR TITLE
General updates, mainly on readers

### PR DESCRIPTION
**Pull request by singletongue**
This PR upgrades the dependencies defined in `requirements.txt` and updates several parts of the `soseki` codes accordingly.
Specifically, the reader is updated so that it can handle inputs produced by a tokenizer other than that of BERT.
Additionally, the experimental results in README.md is also updated.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -142,30 +146,30 @@ $ python evaluate_retriever.py \
...
+# Recall at 100: 0.8501 (3069/3610)
 ```
 
 **5. Train a reader**
 
 ```sh
 $ python train_reader.py \
     --train_file <WORK_DIR>/reader_data/nq_train.jsonl \
-    --dev_file <WORK_DIR>/reader_data/nq_dev.jsonl  \
+    --val_file <WORK_DIR>/reader_data/nq_dev.jsonl \
```
Comment (singletongue) [inline README.md:163]: Changed the fold name `dev` to `val` to coordinate with the naming cnvention of PyTorch Lightning.

File: README.md [hunk 2]
```diff
@@ -142,30 +146,30 @@ $ python evaluate_retriever.py \
...
     --output_dir <WORK_DIR>/reader \
     --train_num_passages 24 \
     --eval_num_passages 100 \
     --max_input_length 256 \
-    --include_title_in_passage \
```
Comment (singletongue) [inline README.md:164]: This option is removed and its behavior is now default.

File: README.md [hunk 3]
```diff
@@ -175,8 +179,7 @@ $ python train_reader.py \
     --max_epochs 20 \
     --gpus 4 \
     --precision 16 \
-    --accelerator ddp \
-    --answer_normalization_type dpr
+    --strategy ddp
```
Comment (singletongue) [inline README.md:182]: The `--accelerator` keyword is deprecated in the latest version of PyTorch Lightning.

File: demo.py [hunk 4]
```diff
@@ -50,32 +53,38 @@ def get_qa_model():
...
+    # Fetch the answer candidates.
     with st.spinner("Computing..."):
-        answer_candidates = qa_model.answer_question(question, num_reading_passages=3)
+        answer_candidates = qa_model.answer_question(question, num_passages_to_read=3)
 
     for i, answer_candidate in enumerate(answer_candidates):
+        input_text = answer_candidate.input_text
         answer_text = answer_candidate.answer_text
-        passage_text = answer_candidate.passage_text
+        answer_start, answer_end = answer_candidate.answer_text_span
```
Comment (singletongue) [inline demo.py:84]: Get the span of the extracted answer so that the answer is highlighted in the passage text.

File: requirements.txt [hunk 5]
```diff
@@ -1,12 +1,12 @@
...
+onnxruntime==1.11.1
+pytorch-lightning==1.6.4
+regex==2022.6.2
 sentencepiece==0.1.96
-streamlit==1.6.0
-torch==1.10.2
-tqdm==4.63.0
-transformers[ja]==4.16.2
+streamlit==1.10.0
+#torch==1.11.0  # You may want to install torch outside this file for the CUDA compatibility.
```
Comment (singletongue) [inline requirements.txt:10]: Since there are different install options for PyTorch depending on environment, we recommend the users to install PyTorch outside the requirements.txt.

File: soseki/utils/data_utils.py [hunk 6]
```diff
@@ -35,129 +31,6 @@ class RetrieverDatasetExample:
     metadata: Optional[Any]
 
 
-class MultipleKeyDataset(Dataset):
```
Comment (singletongue) [inline soseki/utils/data_utils.py:38]: Moved to `soseki/biencoder/modeling.py`.

File: soseki/utils/data_utils.py [hunk 7]
```diff
@@ -35,129 +31,6 @@ class RetrieverDatasetExample:
...
-
-    def __len__(self) -> int:
-        return self.total_dataset_size
-
-    def __getitem__(self, key_idx_tuple: Tuple[str, int]) -> Any:
-        dataset_key, sample_idx = key_idx_tuple
-        return self.dataset_dict[dataset_key][sample_idx]
-
-
-class MultipleKeyDatasetBatchSampler(Sampler[List[int]]):
```
Comment (singletongue) [inline soseki/utils/data_utils.py:59]: Moved to `soseki/biencoder/modeling.py`.

File: soseki/biencoder/modeling.py [hunk 8]
```diff
@@ -2,57 +2,172 @@
...
-    MultipleKeyDatasetBatchSampler,
-    Passage,
-    RetrieverDatasetExample,
-    readitem_json,
-)
+from ..utils.data_utils import Passage, RetrieverDatasetExample, readitem_json
 
 
-class EncoderModel(nn.Module):
+class MultipleKeyDataset(Dataset):
```
Comment (singletongue) [inline soseki/biencoder/modeling.py:26]: Moved from `soseki/utils/data_utils.py`.

File: soseki/biencoder/modeling.py [hunk 9]
```diff
@@ -2,57 +2,172 @@
...
+
+    def __len__(self) -> int:
+        return self.total_dataset_size
+
+    def __getitem__(self, key_idx_tuple: Tuple[str, int]) -> Any:
+        dataset_key, sample_idx = key_idx_tuple
+        return self.dataset_dict[dataset_key][sample_idx]
+
+
+class MultipleKeyDatasetBatchSampler(Sampler[List[int]]):
```
Comment (singletongue) [inline soseki/biencoder/modeling.py:47]: Moved from `soseki/utils/data_utils.py`.

File: soseki/biencoder/modeling.py [hunk 10]
```diff
@@ -2,57 +2,172 @@
...
+
+
+class EncoderModel(nn.Module):
+    def __init__(self, base_model_name: str, pooling_index: int = 0, projection_dim: Optional[int] = None) -> None:
         super().__init__()
         self.encoder = AutoModel.from_pretrained(base_model_name, add_pooling_layer=False, return_dict=False)
-        # For LUKE / mLUKE models which take large amounts of memory for entity embeddings
-        if hasattr(self.encoder, "entity_embeddings"):
-            del self.encoder.entity_embeddings
-
```
Comment (singletongue) [inline soseki/biencoder/modeling.py:43]: This process is no longer needed since `luke-*-lite` models are available.

File: soseki/biencoder/modeling.py [hunk 11]
```diff
@@ -2,57 +2,172 @@
...
         self.pooling_index = pooling_index
 
         if projection_dim is not None:
             self.projection_dense = nn.Linear(self.encoder.config.hidden_size, projection_dim)
             self.projection_layer_norm = torch.nn.LayerNorm(projection_dim, eps=self.encoder.config.layer_norm_eps)
+            self.output_dim = projection_dim
+        else:
+            self.output_dim = self.encoder.config.hidden_size
```
Comment (singletongue) [inline soseki/biencoder/modeling.py:160]: `self.output_dim` is used for checking tensor sizes.

File: soseki/biencoder/modeling.py [hunk 12]
```diff
@@ -2,57 +2,172 @@
...
 import torch.nn.functional as F
 from torch import Tensor
 from torch.optim import AdamW, Optimizer
-from torch.utils.data import Dataset, DataLoader, DistributedSampler
+from torch.utils.data import Dataset, DataLoader, DistributedSampler, Sampler
 from tqdm import tqdm
 from pytorch_lightning.core import LightningModule
-from pytorch_lightning.utilities.distributed import distributed_available, rank_zero_info
+from pytorch_lightning.utilities.distributed import distributed_available
+from pytorch_lightning.utilities.rank_zero import rank_zero_info
```
Comment (singletongue) [inline soseki/biencoder/modeling.py:17]: `rank_zero_info` is moved to `pytorch_lightning.utilities.rank_zero` in the latest version of PyTorch Lightning.

File: soseki/biencoder/modeling.py [hunk 13]
```diff
@@ -432,150 +558,287 @@ def forward(
...
         encoded_question, encoded_passage = batch_parts_outputs
 
+        # Gather tensors distributed to multiple devices.
         if distributed_available():
-            encoded_question = self._gather_distributed_tensors(encoded_question)
-            encoded_passage = self._gather_distributed_tensors(encoded_passage)
+            encoded_question = self.all_gather(encoded_question, sync_grads=True)
+            encoded_question = encoded_question.view(-1, self.question_encoder.output_dim)
+            encoded_passage = self.all_gather(encoded_passage, sync_grads=True)
+            encoded_passage = encoded_passage.view(-1, self.passage_encoder.output_dim)
```
Comment (singletongue) [inline soseki/biencoder/modeling.py]: We now use `LightningModule.all_gather()` function for gathering tensors distributed to multiple GPUs.

PyTorch Lightning uses [`AllGatherGrad`](https://pytorch-lightning.readthedocs.io/en/1.6.4/_modules/pytorch_lightning/utilities/distributed.html#AllGatherGrad) to compute the gradients of the gathered tensors.

I have confirmed with a small dataset that the accumulated loss is almost same comparing to the previous `self._gather_distributed_tensors()` implementation.
スクリーンショット 2022-06-05 14 49 20

File: soseki/reader/modeling.py [hunk 14]
```diff
@@ -20,27 +21,19 @@
...
 
         self.qa_outputs = nn.Linear(self.encoder.config.hidden_size, 2)
         self.qa_classifier = nn.Linear(self.encoder.config.hidden_size, 1)
 
-        self._pooling_index = pooling_index
-
-    def forward(self, input_ids: Tensor, attention_mask: Tensor, token_type_ids: Tensor) -> Tuple[Tensor]:
-        encoder_output = self.encoder(input_ids, attention_mask=attention_mask, token_type_ids=token_type_ids)
+    def forward(self, input_tensors: Dict[str, Tensor]) -> Tuple[Tensor, Tensor, Tensor]:
+        encoder_output = self.encoder(**input_tensors)
```
Comment (singletongue) [inline soseki/reader/modeling.py:33]: The previous model had only accepted the outputs of BERT. Now it accepts tensors that (basically) any Hugging Face model outputs.

File: soseki/reader/modeling.py [hunk 15]
```diff
@@ -733,70 +781,63 @@ def _compute_answer_accuracy(
...
-            for pi, classifier_logit in zip(top_passage_idxs[qi].tolist(), top_classifier_logits[qi].tolist()):
-                answer_spans = self.tokenization._compute_best_answer_spans(
-                    input_ids=input_ids[qi, pi].tolist(),
-                    attention_mask=attention_mask[qi, pi].tolist(),
-                    token_type_ids=token_type_ids[qi, pi].tolist(),
-                    start_logits=start_logits[qi, pi].tolist(),
-                    end_logits=end_logits[qi, pi].tolist(),
-                    num_answer_spans=num_answer_candidates_per_passage,
-                    max_answer_length=max_answer_length,
+    ) -> List[AnswerCandidate]:
```
Comment (singletongue) [inline soseki/reader/modeling.py:790]: `generate_answer_candidates()` now processes one question at a time while it previously processed a batch of multiple questions.

File: soseki/reader/tokenization.py [hunk 16]
```diff
@@ -10,38 +10,26 @@
...
-        passags_text = self._preprocess_passage_text(passags_text)
+        passage_text = self._preprocess_passage_text(passage_text)
 
-        if self.include_title_in_passage:
-            text_a = question
-            text_b = passage_title + self.tokenizer.sep_token + passags_text
-        else:
-            text_a = question + self.tokenizer.sep_token + passage_title
-            text_b = passags_text
+        passage = passage_title + self.tokenizer.sep_token + passage_text
```
Comment (singletongue) [inline soseki/reader/tokenization.py:22]: The previous behavior when `self.include_title_in_passage is True` is now default.

File: soseki/reader/tokenization.py [hunk 17]
```diff
@@ -148,29 +119,30 @@ def _compute_best_answer_spans(
...
+        input_text = self.tokenizer.convert_tokens_to_string(input_tokens)
+        left_input_text = self.tokenizer.convert_tokens_to_string(left_input_tokens)
         answer_text = self.tokenizer.convert_tokens_to_string(answer_tokens)
-        passage_text = self.tokenizer.convert_tokens_to_string(passage_tokens)
 
-        return answer_text, passage_text
+        answer_start = input_text.find(answer_text, len(left_input_text))
+        answer_end = answer_start + len(answer_text)
+
+        return input_text, answer_text, (answer_start, answer_end)
```
Comment (singletongue) [inline soseki/reader/tokenization.py:141]: `_get_input_and_answer_texts()` now returns the start / end positions of the `answer_text` in the `input_text`.
